src/platforms/windows/daemon/windowsfirewall.cpp

```diff
@@ -0,0 +1,874 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "windowsfirewall.h"
+#include "logger.h"
+#include "leakdetector.h"
+#include "../windowscommons.h"
+#include "../../daemon/interfaceconfig.h"
+
+#include <QApplication>
+#include <QObject>
+#include <QFileInfo>
+#include <QNetworkInterface>
+#include <QScopeGuard>
+#include <QHostAddress>
+#include <QtEndian>
+#include <windows.h>
+#include <fwpmu.h>
+#include <stdio.h>
+#include <comdef.h>
+#include <netfw.h>
+#include "winsock.h"
+
+#include <initguid.h>
+#include <guiddef.h>
+#include <qaccessible.h>
+
+
+// ID for the Firewall Sublayer
+DEFINE_GUID(ST_FW_WINFW_BASELINE_SUBLAYER_KEY,
+    0xc78056ff, 0x2bc1, 0x4211, 0xaa, 0xdd, 0x7f, 0x35, 0x8d, 0xef, 0x20, 0x2d);
+// ID for the Mullvad Split-Tunnel Sublayer Provider
+DEFINE_GUID(ST_FW_PROVIDER_KEY,
+    0xe2c114ee, 0xf32a, 0x4264, 0xa6, 0xcb, 0x3f, 0xa7, 0x99, 0x63, 0x56, 0xd9);
+
+
+namespace {
+Logger logger(LOG_WINDOWS, "WindowsFirewall");
+WindowsFirewall* s_instance = nullptr;
+
+// Note Filter Weight may be between 0-15!
+constexpr uint8_t LOW_WEIGHT = 0;
+constexpr uint8_t MED_WEIGHT = 7;
+constexpr uint8_t HIGH_WEIGHT =13;
+constexpr uint8_t MAX_WEIGHT =15;
+}
+
+WindowsFirewall* WindowsFirewall::instance(){
+    if(s_instance == nullptr){
+        s_instance = new WindowsFirewall(qApp);
+    }
+    return s_instance;
+}
+
+WindowsFirewall::WindowsFirewall(QObject* parent) : QObject(parent) {
+  MVPN_COUNT_CTOR(WindowsFirewall);
+  Q_ASSERT(s_instance == nullptr);
+
+  HANDLE    engineHandle = NULL;
+  DWORD    result = ERROR_SUCCESS;
+  // Use dynamic sessions for efficiency and safety:
+  //  -> Filtering policy objects are deleted even when the application crashes/ deamon goes down
+  FWPM_SESSION0 session;
+  memset(&session, 0, sizeof(session));
+  session.flags = FWPM_SESSION_FLAG_DYNAMIC;
+
+  logger.log() << ("Opening the filter engine.");
+
+  result = FwpmEngineOpen0(
+      NULL,
+      RPC_C_AUTHN_WINNT,
+      NULL,
+      &session,
+      &engineHandle );
+
+  if (result != ERROR_SUCCESS){
+    WindowsCommons::windowsLog("FwpmEngineOpen0 failed");
+    return;
+  }
+  logger.log() << ("Filter engine opened successfully.");
+  m_sessionHandle = engineHandle;
+
+
+}
+
+WindowsFirewall::~WindowsFirewall() {
+  MVPN_COUNT_DTOR(WindowsFirewall);
+  CloseHandle(m_sessionHandle);
```

Comment:
Probably it's not needed, but:
`if (m_sessionHandle != INVALID_HANDLE_VALUE) {` ?